HCAP-1039: view cohort details page

alva-fresh opened the pull request:
https://freshworks.atlassian.net/browse/HCAP-1039

Creating a new `CohortDetails` page to show
- Start date
- End date
- Capacity
- Availability
- Number of unsuccessful participants

<img width="650" alt="Screen Shot 2022-09-07 at 4 06 06 PM" src="https://user-images.githubusercontent.com/64768811/188998649-eddef89a-fcdf-4678-91a7-70c4835cc4d2.png">


There's a separate ticket to display a list of participants attached to this cohort --> https://freshworks.atlassian.net/browse/HCAP-1056

https://www.figma.com/proto/35c5C0DyTEOxJ4N4DIaADL/HCAP-Design?node-id=2441%3A37315&viewport=241%2C48%2C0.37&scaling=min-zoom&starting-point-node-id=2441%3A37147

taraepp reviewed (commented):
I have a couple of questions about the BE but overall looks good to me

arranfw commented on client/src/pages/private/CohortDetails.js:
```diff
@@ -0,0 +1,126 @@
+    );
+    let participantCount = 0;
+    for (const participantId of cohortParticipantIds) {
+      const postHireData = await fetchParticipantPostHireStatus({ id: participantId });
```
I'd like to see this done in a single request or added to the original `fetchCohortDetails` request, iterating requests for participants isn't efficient

alva-fresh replied:
Yes, you're right. I'll have it in separate request for HCAP-1056

arranfw commented on client/src/pages/private/CohortDetails.js:
```diff
@@ -0,0 +1,126 @@
+      fetchUnsuccessfulParticipantsNumber();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [cohort, unsuccessfulParticipantsNumber]);
```
If you leave this dep array empty you can omit the conditionals

Is there any case where you want this effect to run other than on mount?

alva-fresh replied:
Actually no, this array can be empty

arranfw commented on client/src/pages/private/CohortDetails.js:
```diff
@@ -0,0 +1,126 @@
+  const [unsuccessfulParticipantsNumber, setUnsuccessfulParticipantsNumber] = useState(-1);
+  const { openToast } = useToast();
+
+  const getTotalParticipantsNumber = () => {
+    return cohort ? cohort.cohort_size : 0;
+  };
+
+  const getAvailableParticipantsNumber = () => {
+    if (!cohort) return 0;
+    const totalNumber = cohort.cohort_size;
+    const assignedParticipantsNumber = cohort.participants?.length;
+    return totalNumber - assignedParticipantsNumber;
+  };
```
```suggestion
  const cohortSize = cohort?.cohort_size || 0;
  const assignedParticipants = cohort?.participants.length || 0;
  const availableCohortSeats = cohortSize - assignedParticipants;
```

These functions can be replaced by a few variables, I also adjusted their names to be a bit more reflective of their data

The benefit here is that these variables can be used freely by other parts of the component rather than being nested in a function, as well as just being simpler than creating a function

arranfw commented on client/src/pages/private/CohortTable.js:
```diff
@@ -33,11 +37,24 @@ export default ({ cohorts, editCohortAction }) => {
+          <Link
+            component='button'
+            variant='body2'
+            onClick={() => {
```
Could this be extracted to a fn `handleCohortNav`?

alva-fresh replied:
Good point!

arranfw commented on client/src/pages/private/PSIView.js:
```diff
@@ -213,35 +213,13 @@ export default () => {
-                  */
-                  text='Download participants attending PSI Report'
+                  startIcon={<GetAppIcon />}
+                  text='Download PSI list (.csv)'
```
I didn't see this in any AC for HCAP-1039

alva-fresh replied:
This is just to match the mockups we have

<img width="500" alt="Screen Shot 2022-09-08 at 2 29 20 PM" src="https://user-images.githubusercontent.com/64768811/189229483-36191abf-85e5-4348-8eb5-dc73028fd0d7.png">

arranfw commented on client/src/pages/private/PSIViewDetails.js:
```diff
@@ -208,13 +208,11 @@ export default ({ match }) => {
               <Box pb={2}>
                 <Typography variant='body1'>
-                  <Link href={routes.PSIView}>PSI</Link> / {psi.instituteName}
+                  <Link href={routes.PSIView}>Manage PSI</Link> / PSI Details
```
Same here, why was this change made?

arranfw commented on client/src/services/psi-cohort.js:
```diff
@@ -101,6 +101,22 @@ export const fetchCohorts = async ({ psiId }) => {
 };
 
+export const fetchCohort = async ({ cohortId }) => {
+  const response = await fetch(`${API_URL}/api/v1/cohorts/${cohortId}/participants`, {
```
This route structure would make me expect a list of participants, not cohort details, why not just 

`/cohorts/${cohortId}`

arranfw reviewed (changes requested):
I've added some comments about changes to the FE component, API structure and a few questions

alva-fresh commented on server/routes/cohorts.js:
```diff
@@ -70,6 +71,28 @@ router.get(
+    const [cohort] = await getCohortWithParticipants(id);
+    if (cohort === undefined) {
+      return res.status(401).send({ message: 'You do not have permission to view this record' });
+    }
```
I'll update that 👍

alva-fresh replied:
That's right, it returns an empty array. I'll remove this check

alva-fresh commented:
@arranfw @taraepp I addressed your comments in https://github.com/bcgov/hcap/pull/835/files - it also includes some refactoring to this request as well as adding participants table. I'm sorry about the inconsistency!
Please take a look 👀

fw-noel reviewed (changes requested):
Please see my last comment

taraepp reviewed (approved):
LGTM!
